File: 1675-minimize-deviation-in-array/1675-minimize-deviation-in-array.kt
class Solution {
    fun minimumDeviation(nums: IntArray): Int {
        val set: TreeSet<Int> = TreeSet()
        
        for (i in nums) {
            when (i % 2) {
                1 -> set.add(i * 2)
                else -> set.add(i)
            }
        }
        
        var result: Int = set.last() - set.first()
        while (set.last() % 2 == 0) {
            val current = set.last()
            set.remove(current)
            set.add(current / 2)
            result = minOf(result, set.last() - set.first())
        }
        
        return result
    }
}

File: 1675-minimize-deviation-in-array/1675-minimize-deviation-in-array.cpp
class Solution {
public:
    int minimumDeviation(vector<int>& nums) {
        set<int> set;
        
        for (auto &i : nums) {
            if (i % 2 == 1) set.insert(i * 2);
            else set.insert(i);
        }
        
        int result = *set.rbegin() - *set.begin();
        while (*set.rbegin() % 2 == 0) {
            int current = *set.rbegin();
            set.erase(current);
            set.insert(current / 2);
            result = min(result, *set.rbegin() - *set.begin());
        }
        
        return result;
    }
};